airbyte-integrations/connectors/destination-langchain/destination_langchain/indexer.py

```diff
@@ -0,0 +1,137 @@
+#
+# Copyright (c) 2023 Airbyte, Inc., all rights reserved.
+#
+
+import itertools
+import os
+import uuid
+from abc import ABC, abstractmethod
+from typing import Any, List, Optional
+
+import pinecone
+from airbyte_cdk.models import ConfiguredAirbyteCatalog
+from airbyte_cdk.models.airbyte_protocol import DestinationSyncMode
+from destination_langchain.config import DocArrayHnswSearchIndexingModel, PineconeIndexingModel
+from destination_langchain.document_processor import METADATA_NATURAL_ID_FIELD, METADATA_STREAM_FIELD
+from destination_langchain.embedder import Embedder
+from destination_langchain.measure_time import measure_time
+from langchain.document_loaders.base import Document
+from langchain.vectorstores.docarray import DocArrayHnswSearch
+
+
+class Indexer(ABC):
+    def __init__(self, config: Any, embedder: Embedder):
+        self.config = config
+        self.embedder = embedder
+        pass
+
+    def pre_sync(self, catalog: ConfiguredAirbyteCatalog):
+        pass
+
+    def post_sync(self):
+        pass
+
+    @abstractmethod
+    def index(self, document_chunks: List[Document], delete_ids: List[str]):
+        pass
+
+    @abstractmethod
+    def check(self) -> Optional[str]:
+        pass
+
+    @property
+    def max_metadata_size(self) -> Optional[int]:
+        return None
+
+
+def chunks(iterable, batch_size):
+    """A helper function to break an iterable into chunks of size batch_size."""
+    it = iter(iterable)
+    chunk = tuple(itertools.islice(it, batch_size))
+    while chunk:
+        yield chunk
+        chunk = tuple(itertools.islice(it, batch_size))
+
+
+class PineconeIndexer(Indexer):
+    config: PineconeIndexingModel
+
+    def __init__(self, config: PineconeIndexingModel, embedder: Embedder):
+        super().__init__(config, embedder)
+        pinecone.init(api_key=config.pinecone_key, environment=config.pinecone_environment, threaded=True)
+        self.pinecone_index = pinecone.Index(config.index, pool_threads=10)
+        self.embed_fn = measure_time(self.embedder.langchain_embeddings.embed_documents)
+
+    def pre_sync(self, catalog: ConfiguredAirbyteCatalog):
+        for stream in catalog.streams:
+            if stream.destination_sync_mode == DestinationSyncMode.overwrite:
+                self.pinecone_index.delete(filter={METADATA_STREAM_FIELD: stream.stream.name})
+
+    def post_sync(self):
+        self.embed_fn._print_stats()
```

Comment:
do we want to print or send as logs? I'm not sure this will make it to the connector logs

Reply:
They show up, but it's probably cleaner to yield them as log messages. Changed that